neps/nep-0364.md

```diff
@@ -0,0 +1,171 @@
+---
+NEP: 364
+Title: Efficient signature verification and hashing precompile functions
+Author: Blas Rodriguez Irizar <rodrigblas@gmail.com>
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/364
+Status: Draft
+Type: Runtime Spec
+Category: Contract
+Created: 15-Jun-2022
+---
+
+## Summary
+
+This NEP introduces the request of adding into the NEAR runtime a set of pre-compile
+functions that can help IBC compatible light clients run on-chain.
+
+## Motivation
+
+Signature verification and hashing are ubiquitous operations in light clients,
+especially in PoS consensus mechanisms. Based on the Polkadot's consensus mechanism
+there will be a need for verification of ~200 signatures every minute
+(Polkadot’s authority set is ~300 signers and it may be incresed in the future: https://polkadot.polkassembly.io/referendum/16).
+
+Therefore, a mechanism to perform these operations in a cost-effective way
+in terms of gas and speed would be highly beneficial to have. Currently, NEAR does not have any native signature verification toolbox.
+This implies that a light client operating insid NEAR will have to import a library
+compiled to WASM as mentioned in [Zulip](https://near.zulipchat.com/#narrow/stream/295302-general/topic/light_client).
+
+Polkadot uses [three different cryptographic schemes](https://wiki.polkadot.network/docs/learn-keys)
+for its keys/accounts, which also translates into different signatures types:
+
+- The vanilla ed25519 implementation using Schnorr signatures.
+- The Schnorrkel/Ristretto sr25519 variant using Schnorr signatures.
+- ECDSA signatures on secp256k1
+
+## Rationale and alternatives
+
+Add three signature verification signatures functions into the runtime as host functions.
+
+- ED25519 signature verification function
+- SR25519 signature verification function
+- ECDSA signature verification function
+
+using `ed25519_dalek`, `schnorrkel` and `secp256k1` crates into NEAR runtime as pre-compiled functions.
+
+Note: seems like `ecdsa_verify` is implemented, but called `erecover` in https://github.com/near/nearcore/blob/master/runtime/near-vm-logic/src/logic.rs#L1049
+Upon confirmation, we can proceed to remove it.
+
+Benchmarks were run using a signature verifier smart contract on-chain importing the aforementioned functions from
+widespread used crypto Rust crates. The biggest pitfall of that solution
```

Comment:
What solution is “that solution” referring to here? I would imagine computing these signatures in WASM? Please make the wording more precise.